File: Collections/NullAcceptingDictionary.vb
Namespace Common

    Public Class NullAcceptingDictionary

        Public Shared Function Create(Of TKey As Class, TValue)(ByVal Dic As IDictionary(Of TKey, TValue)) As NullAcceptingDictionary(Of TKey, TValue)
            Return New NullAcceptingDictionary(Of TKey, TValue)(Dic)
        End Function

    End Class

    Public Class NullAcceptingDictionary(Of TKey As Class, TValue)
        Inherits BaseDictionary(Of TKey, TValue)

        Public Sub New(ByVal Dic As IDictionary(Of TKey, TValue))
            Me.Dic = Dic
            Me.KeysCollection = New MergedCollection(Of TKey)(New TKey(0) {Nothing}, Me.Dic.Keys)
            Me.ValuesCollection = New MergedCollection(Of TValue)(Me.NullValue, Me.Dic.Values)
        End Sub

        Public Sub New()
            Me.New(New Dictionary(Of TKey, TValue)())
        End Sub

        Protected Overrides ReadOnly Property IsReadOnly As Boolean
            Get
                Return Me.Dic.IsReadOnly
            End Get
        End Property

        Public Overrides ReadOnly Property Count As Integer
            Get
                If Me.HasNullValue Then
                    Return Me.Dic.Count + 1
                End If
                Return Me.Dic.Count
            End Get
        End Property

        Default Public Overrides Property Item(key As TKey) As TValue
            Get
                If key Is Nothing Then
                    If Not Me.HasNullValue Then
                        Throw New KeyNotFoundException()
                    End If
                    Return Me.NullValue(0)
                End If
                Return Me.Dic.Item(key)
            End Get
            Set(value As TValue)
                If key Is Nothing Then
                    If Me.IsReadOnly Then
                        Throw New NotSupportedException("Collection is read only.")
                    End If
                    Me.NullValue(0) = value
                    Me.HasNullValue = True
                    Exit Property
                End If
                Me.Dic.Item(key) = value
            End Set
        End Property

        Public Overrides ReadOnly Property Keys As ICollection(Of TKey)
            Get
                If Not Me.HasNullValue Then
                    Return Me.Dic.Keys
                End If
                Return Me.KeysCollection
            End Get
        End Property

        Public Overrides ReadOnly Property Values As ICollection(Of TValue)
            Get
                If Not Me.HasNullValue Then
                    Return Me.Dic.Values
                End If
                Return Me.ValuesCollection
            End Get
        End Property

        Protected Overrides Sub CopyTo(ByVal Array As Array, ByVal ArrayIndex As Integer)
            If Me.HasNullValue Then
                Array.SetValue(New KeyValuePair(Of TKey, TValue)(Nothing, Me.NullValue(0)), ArrayIndex)
                ArrayIndex += 1
            End If

            Dim Dic = TryCast(Me.Dic, IDictionary)
            If Dic IsNot Nothing Then
                Dic.CopyTo(Array, ArrayIndex)
                Exit Sub
            End If

            For Each KV In Dic
                Array.SetValue(KV, ArrayIndex)
                ArrayIndex += 1
            Next
        End Sub

        Public Overrides Sub Add(key As TKey, value As TValue)
            If key Is Nothing Then
                If Me.IsReadOnly Then
                    Throw New NotSupportedException()
                End If
                Me.NullValue(0) = value
                Me.HasNullValue = True
                Exit Sub
            End If
            Me.Dic.Add(key, value)
        End Sub

        Public Overrides Sub Clear()
            If Me.IsReadOnly Then
                Throw New NotSupportedException()
            End If
            Me.NullValue(0) = Nothing
            Me.HasNullValue = False
            Me.Dic.Clear()
        End Sub

        Public Overrides Function ContainsKey(key As TKey) As Boolean
            If key Is Nothing Then
                Return Me.HasNullValue
            End If
            Return Me.Dic.ContainsKey(key)
        End Function

        Public Overrides Function Remove(key As TKey) As Boolean
            If key Is Nothing Then
                If Me.IsReadOnly Then
                    Throw New NotSupportedException()
                End If
                If Not Me.HasNullValue Then
                    Return False
                End If
                Me.NullValue(0) = Nothing
                Me.HasNullValue = False
                Return True
            End If
            Return Me.Dic.Remove(key)
        End Function

        Public Overrides Function TryGetValue(key As TKey, ByRef value As TValue) As Boolean
            If key Is Nothing Then
                If Me.HasNullValue Then
                    value = Me.NullValue(0)
                    Return True
                End If
                Return False
            End If

            Return Me.Dic.TryGetValue(key, value)
        End Function

        Private Iterator Function GetEnumeratorWithNull() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Yield New KeyValuePair(Of TKey, TValue)(Nothing, Me.NullValue(0))
            For Each KV In Me.Dic
                Yield KV
            Next
        End Function

        Public Function GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            If Me.HasNullValue Then
                Return Me.GetEnumeratorWithNull()
            End If
            Return Me.Dic.GetEnumerator()
        End Function

        Protected Overrides Function IEnumerator_1_GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Return Me.GetEnumerator()
        End Function

        Private ReadOnly Dic As IDictionary(Of TKey, TValue)
        Private HasNullValue As Boolean
        Private ReadOnly NullValue As TValue() = New TValue(0) {}

        Private ReadOnly KeysCollection As MergedCollection(Of TKey)
        Private ReadOnly ValuesCollection As MergedCollection(Of TValue)

    End Class

End Namespace

File: Collections/NullAcceptingDictionary.cs
using System.Collections.Generic;
using System.Collections;
using System;

namespace Ks.Common
{
    public class NullAcceptingDictionary
    {
        public static NullAcceptingDictionary<TKey, TValue> Create<TKey, TValue>(IDictionary<TKey, TValue> Dic) where TKey : class
        {
            return new NullAcceptingDictionary<TKey, TValue>(Dic);
        }
    }

    public class NullAcceptingDictionary<TKey, TValue> : BaseDictionary<TKey, TValue> where TKey : class
    {
        public NullAcceptingDictionary(IDictionary<TKey, TValue> Dic)
        {
            this.Dic = Dic;
            this.KeysCollection = new MergedCollection<TKey>(new TKey[1] { null }, this.Dic.Keys);
            this.ValuesCollection = new MergedCollection<TValue>(this.NullValue, this.Dic.Values);
        }

        public NullAcceptingDictionary() : this(new Dictionary<TKey, TValue>())
        {
        }

        protected override bool IsReadOnly
        {
            get
            {
                return this.Dic.IsReadOnly;
            }
        }

        public override int Count
        {
            get
            {
                if (this.HasNullValue)
                    return this.Dic.Count + 1;
                return this.Dic.Count;
            }
        }

        public override TValue this[TKey key]
        {
            get
            {
                if (key == null)
                {
                    if (!this.HasNullValue)
                        throw new KeyNotFoundException();
                    return this.NullValue[0];
                }
                return this.Dic[key];
            }
            set
            {
                if (key == null)
                {
                    if (this.IsReadOnly)
                        throw new NotSupportedException("Collection is read only.");
                    this.NullValue[0] = value;
                    this.HasNullValue = true;
                    return;
                }
                this.Dic[key] = value;
            }
        }

        public override ICollection<TKey> Keys
        {
            get
            {
                if (!this.HasNullValue)
                    return this.Dic.Keys;
                return this.KeysCollection;
            }
        }

        public override ICollection<TValue> Values
        {
            get
            {
                if (!this.HasNullValue)
                    return this.Dic.Values;
                return this.ValuesCollection;
            }
        }

        protected override void CopyTo(Array Array, int ArrayIndex)
        {
            if (this.HasNullValue)
            {
                Array.SetValue(new KeyValuePair<TKey, TValue>(null, this.NullValue[0]), ArrayIndex);
                ArrayIndex += 1;
            }

            var Dic = this.Dic as IDictionary;
            if (Dic != null)
            {
                Dic.CopyTo(Array, ArrayIndex);
                return;
            }

            foreach (var KV in Dic)
            {
                Array.SetValue(KV, ArrayIndex);
                ArrayIndex += 1;
            }
        }

        public override void Add(TKey key, TValue value)
        {
            if (key == null)
            {
                if (this.IsReadOnly)
                    throw new NotSupportedException();
                this.NullValue[0] = value;
                this.HasNullValue = true;
                return;
            }
            this.Dic.Add(key, value);
        }

        public override void Clear()
        {
            if (this.IsReadOnly)
                throw new NotSupportedException();
            this.NullValue[0] = default;
            this.HasNullValue = false;
            this.Dic.Clear();
        }

        public override bool ContainsKey(TKey key)
        {
            if (key == null)
                return this.HasNullValue;
            return this.Dic.ContainsKey(key);
        }

        public override bool Remove(TKey key)
        {
            if (key == null)
            {
                if (this.IsReadOnly)
                    throw new NotSupportedException();
                if (!this.HasNullValue)
                    return false;
                this.NullValue[0] = default;
                this.HasNullValue = false;
                return true;
            }
            return this.Dic.Remove(key);
        }

        public override bool TryGetValue(TKey key, out TValue value)
        {
            if (key == null)
            {
                if (this.HasNullValue)
                {
                    value = this.NullValue[0];
                    return true;
                }
                value = default;
                return false;
            }

            return this.Dic.TryGetValue(key, out value);
        }

        private IEnumerator<KeyValuePair<TKey, TValue>> GetEnumeratorWithNull()
        {
            yield return new KeyValuePair<TKey, TValue>(null, this.NullValue[0]);
            foreach (var KV in this.Dic)
                yield return KV;
        }

        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
        {
            if (this.HasNullValue)
                return this.GetEnumeratorWithNull();
            return this.Dic.GetEnumerator();
        }

        protected override IEnumerator<KeyValuePair<TKey, TValue>> _GetEnumerator()
        {
            return this.GetEnumerator();
        }

        private readonly IDictionary<TKey, TValue> Dic;
        private bool HasNullValue;
        private readonly TValue[] NullValue = new TValue[1];

        private readonly MergedCollection<TKey> KeysCollection;
        private readonly MergedCollection<TValue> ValuesCollection;
    }
}
